modules/images/image-loading-optimization/class-ilo-url-metrics-group-collection.php

```diff
@@ -0,0 +1,241 @@
+<?php
+/**
+ * Image Loading Optimization: ILO_URL_Metrics_Group_Collection class
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+/**
+ * Collection of URL groups according to the breakpoints.
+ *
+ * @since n.e.x.t
+ * @access private
+ */
+final class ILO_URL_Metrics_Group_Collection {
+
+	/**
+	 * URL metrics groups.
+	 *
+	 * The number of groups corresponds to one greater than the number of
+	 * breakpoints. This is because breakpoints are the dividing line between
+	 * the groups of URL metrics with specific viewport widths. This extends
+	 * even to when there are zero breakpoints: there will still be one group
+	 * in this case, in which every single URL metric is added.
+	 *
+	 * @var ILO_URL_Metrics_Group[]
+	 * @phpstan-var non-empty-array<ILO_URL_Metrics_Group>
+	 */
+	private $groups;
+
+	/**
+	 * Breakpoints in max widths.
+	 *
+	 * @var int[]
+	 * @phpstan-var int<1, max>[]
+	 */
+	private $breakpoints;
```

Comment:
So this is for instance 480, 800, and ... (what's used for the maximum?), which would mean the groups are:
* 0-480
* 481-800
* 800-infinite

Do I understand that correctly?

Reply:
If two breakpoints were provided: 480 and 800, then there would be 3 groups with viewport ranges:

1. 0-600
2. 601-800
3. 801-∞

A breakpoint must be greater than zero, because it is not possible to have a viewport that is zero pixels wide. A breakpoint wouldn't in practice be the maximum because that would mean there could be a viewport at maximum + 1. In this way, the inclusion of `max` in the typing is not accurate. It should rather be `int<1, max-1>[]` but I don't think that's supported.

Reply:
Further improved typing in 34e3d062 to clarify the range of values allowed for breakpoints.